File: Source/CDKStack.m
//
//  CDKStack.m
//  CoreDataKit
//
//  Created by Felix Gabel on 13/11/14.
//  Copyright (c) 2014 Felix Gabel. All rights reserved.
//

#import "CDKStack.h"
#import "NSURL+CoreDataKit.h"


@implementation CDKStack

- (instancetype)initWithModel:(NSManagedObjectModel *)model {
	if ((self = [super init])) {
		_coordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:model];
		_mainContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
		_mainContext.persistentStoreCoordinator = _coordinator;
		_model = model;
	}
	return self;
}

- (instancetype)init {
	NSManagedObjectModel *model = [NSManagedObjectModel mergedModelFromBundles:nil];
	return [self initWithModel:model];
}


#pragma mark -

- (NSPersistentStore *)addStoreAtURL:(NSURL *)url withOptions:(NSDictionary *)options configuration:(NSString *)configuration {
	NSFileManager *fileManager = [NSFileManager defaultManager];
	NSURL *directoryURL = url.URLByDeletingLastPathComponent;
	
	if ([fileManager createDirectoryAtURL:directoryURL withIntermediateDirectories:YES attributes:nil error:nil]) {
		return [self.coordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:configuration URL:url options:options error:nil];
	}
	return nil;
}

- (NSPersistentStore *)addStoreAtURL:(NSURL *)url withOptions:(NSDictionary *)options {
	return [self addStoreAtURL:url withOptions:options configuration:nil];
}

- (NSPersistentStore *)addStoreAtURL:(NSURL *)url {
	NSDictionary *options = @{NSMigratePersistentStoresAutomaticallyOption:@YES, NSInferMappingModelAutomaticallyOption:@YES};
	return [self addStoreAtURL:url withOptions:options];
}

- (NSPersistentStore *)addStore {
	NSURL *url = [NSURL defaultStoreURL];
	return [self addStoreAtURL:url];
}

@end

File: Source/CDKStack.swift
//
//  CDKStack.swift
//  CoreDataKit
//
//  Created by Felix Gabel on 21/07/14.
//  Copyright (c) 2014 Felix Gabel. All rights reserved.
//

import CoreData


public class CDKStack {

	public let model:NSManagedObjectModel
	public let mainContext:NSManagedObjectContext
	public let coordinator:NSPersistentStoreCoordinator
	
	
	public init(_ model:NSManagedObjectModel = NSManagedObjectModel.mergedModelFromBundles(nil)) {
		self.coordinator = NSPersistentStoreCoordinator(managedObjectModel:model)
		self.mainContext = NSManagedObjectContext(concurrencyType:.MainQueueConcurrencyType)
		self.mainContext.persistentStoreCoordinator = self.coordinator
		self.model = model
	}
	

	/// Adds a new persistent store with specific information and returns the new store
	public func addStore(_ info:CDKStoreInfo = CDKStoreInfo()) -> NSPersistentStore {
		let fileManager = NSFileManager.defaultManager()
		
		fileManager.createDirectoryAtURL(info.directoryURL, withIntermediateDirectories:true, attributes:nil, error:nil)
		return self.coordinator.addPersistentStoreWithType(info.type, configuration:info.configuration, URL:info.URL, options:info.options, error:nil)!
	}
}
